lang/src/org/partiql/lang/eval/visitors/AggregationVisitorTransform.kt

```diff
@@ -0,0 +1,364 @@
+
+import com.amazon.ionelement.api.emptyMetaContainer
+import org.partiql.lang.ast.UNKNOWN_SOURCE_LOCATION
+import org.partiql.lang.ast.passes.SemanticException
+import org.partiql.lang.ast.passes.SemanticProblemDetails
+import org.partiql.lang.domains.PartiqlAst
+import org.partiql.lang.domains.toBindingCase
+import org.partiql.lang.errors.ErrorCode
+import org.partiql.lang.errors.Problem
+import org.partiql.lang.errors.Property
+import org.partiql.lang.eval.BindingName
+import org.partiql.lang.eval.EvaluationException
+import org.partiql.lang.eval.errorContextFrom
+import org.partiql.lang.eval.extractColumnAlias
+import org.partiql.lang.eval.sourceLocationMeta
+
+/**
+ * This VisitorTransform:
+ * - adds unique aliases to each group key
+ * - replaces group key references (ids) with their unique name
+ * - replaces group key references (ids) with the expression that they represent (if within an aggregation function)
+ * - handles scoping by keeping a context stack
+ * - converts project star into the aggregation outputs (group keys and group as)
+ * - throws errors on projection items that reference variables that are not part of the aggregation operator output
+ *
+ * This VisitorTransform is specifically used by the planner implementation.
+ *
+ * Turns:
+ *
+ * ```
+ * SELECT
+ *     k AS group_key,
+ *     SUM(k) AS the_sum,
+ *     g AS the_input,
+ *     (
+ *         SELECT k + SUM(k), SUM(h)
+ *         FROM t2
+ *         GROUP BY t2.a AS h
+ *     ) AS projection_query
+ * FROM t1
+ * GROUP BY t1.a AS k, t1.b AS h
+ * GROUP AS g
+ * HAVING k + SUM(k) > 0
+ * ORDER BY
+ *     k + SUM(k)
+ *     AND
+ *     (SELECT k + SUM(k), SUM(h) FROM t2)
+ * ```
+ *
+ * Into:
+ *
+ * ```
+ * SELECT
+ *     $__partiql__group_by_0_0 AS group_key,
+ *     SUM(t1.a) AS the_sum,
+ *     "g" AS the_input,
+ *     (
+ *         SELECT $__partiql__group_by_0_0 + SUM($__partiql__group_by_0_0), SUM($__partiql__group_by_1_0)
+ *         FROM t2
+ *         GROUP BY t2.a AS $__partiql__group_by_1_0
+ *     ) AS projection_query
+ * FROM t1
+ * GROUP BY t1.a AS $__partiql__group_by_0_0, t1.b AS $__partiql__group_by_0_0
+ * GROUP AS g
+ * HAVING $__partiql__group_by_0_0 + SUM(t1.a) > 0
+ * ORDER BY
+ *     $__partiql__group_by_0_0 + SUM(t1.a)
+ *     AND
+ *     (SELECT $__partiql__group_by_0_0 + SUM($__partiql__group_by_0_0), SUM($__partiql__group_by_0_1) FROM t2)
+ * ```
+ *
+ */
+internal class AggregationVisitorTransform(
+    private val contextStack: MutableList<VisitorContext> = mutableListOf()
+) : VisitorTransformBase() {
+
+    private val itemTransform = GroupKeyReferenceTransformer(contextStack)
+
+    companion object {
+        internal const val GROUP_PREFIX = "\$__partiql__group_by_"
+        internal const val GROUP_DELIMITER = "_"
+        internal fun uniqueAlias(level: Int, index: Int) = "$GROUP_PREFIX$level$GROUP_DELIMITER$index"
+    }
+
+    /**
+     * Every ExprSelect adds to the [contextStack] during [transformExprSelect_group].
```

Comment:
Could you elaborate a little more on this comments.

The logic seems to be we essentially want to add to the `contextStack` upon entering a subquery, and corresponding level remove the context upon leaving a subquery.

Reply:
The transformExprSelectEvaluationOrder transforms each clause based on evaluation order. As part of that, it goes into Select.Group (even if it's null). So, since we guarantee that we will go into transformExprSelect_group and we have all of the grouping information there, we opt to add to the context there. Then, I can pop the context off after this exprSelect. And, the only clauses the care about the context are clauses after the Group.

LMK if that makes sense. But essentially, we add to the ContextStack in the TransformExprSelect_group always.

Reply:
Added a comment for this.